rust/rust.cc

```diff
@@ -0,0 +1,323 @@
+            auto loc = ctx->getPipLocation(uh);
+            src_loc.x += loc.x;
+            src_loc.y += loc.y;
+            uh_pips++;
+        }
+        if (uh_pips > 1) {
+            src_loc.x /= uh_pips;
+            src_loc.y /= uh_pips;
+        }
+
+        auto dh_pips = 0;
+        for (auto dh : ctx->getPipsDownhill(ctx->getPipDstWire(pip))) {
+            auto loc = ctx->getPipLocation(dh);
+            dst_loc.x += loc.x;
+            dst_loc.y += loc.y;
+            dh_pips++;
+        }
+        if (dh_pips > 1) {
+            dst_loc.x /= dh_pips;
+            dst_loc.y /= dh_pips;
+        }
+
+        dst_loc.x -= src_loc.x;
+        dst_loc.y -= src_loc.y;
+        return dst_loc;
+    }
+
+    uint64_t npnr_context_get_pips_leak(const Context *const ctx, PipId **pips) {
+        auto pip_vec = std::vector<PipId>{};
+        for (auto pip : ctx->getPips()) {
+            pip_vec.push_back(pip);
+        }
+        pip_vec.shrink_to_fit();
+        auto size = pip_vec.size();
+        *pips = pip_vec.data();
+        auto dummy = std::vector<PipId>{};
+        // Yes, by memcpying over `pip_vec` we leak memory.
+        std::memcpy(&pip_vec, &dummy, sizeof(dummy));
+        return size;
+    }
+
+    uint64_t npnr_context_get_wires_leak(const Context *const ctx, WireId **wires) {
+        auto wire_vec = std::vector<WireId>{};
+        for (auto wire : ctx->getWires()) {
+            wire_vec.push_back(wire);
+        }
+        wire_vec.shrink_to_fit();
+        auto size = wire_vec.size();
+        *wires = wire_vec.data();
+        auto dummy = std::vector<WireId>{};
+        // Yes, by memcpying over `wire_vec` we leak memory.
+        std::memcpy(&wire_vec, &dummy, sizeof(dummy));
+        return size;
+    }
+
+    void npnr_context_check(const Context *const ctx) { ctx->check(); }
+    bool npnr_context_debug(const Context *const ctx) { return ctx->debug; }
+    int npnr_context_id(const Context *const ctx, const char *const str) { return ctx->id(str).hash(); }
+    const char *npnr_context_name_of(const Context *const ctx, IdString str) { return ctx->nameOf(str); }
+    const char *npnr_context_name_of_pip(const Context *const ctx, uint64_t pip) { return ctx->nameOfPip(unwrap_pip(pip)); }
+    const char *npnr_context_name_of_wire(const Context *const ctx, uint64_t wire) { return ctx->nameOfWire(unwrap_wire(wire)); }
+    bool npnr_context_verbose(const Context *const ctx) { return ctx->verbose; }
+
+    uint64_t npnr_context_get_netinfo_source_wire(const Context *const ctx, const NetInfo *const net) { return wrap(ctx->getNetinfoSourceWire(net)); }
+    uint64_t npnr_context_get_netinfo_sink_wire(const Context *const ctx, const NetInfo *const net, const PortRef *const sink, uint32_t n) { return wrap(ctx->getNetinfoSinkWire(net, *sink, n)); }
+
+    uint32_t npnr_context_nets_leak(const Context *const ctx, int **names, NetInfo ***nets) {
+        auto name_vec = std::vector<int>{};
```

Comment:
Similar comment as to the `WireId` approach, except here you wouldn't even need a first loop to get the size because you could just call `ctx->nets.size()`...